File: Design_Hashset.java
// Time Complexity : O(1)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
//Approach : For this problem we will create array in array. The primary array will have the index corresponding to the first hash function.
// and the nested array which will be boolean will have index corresponding to the second hash function. hash functions will be such that it will avoid all the collisions.
//at the 0 index of primary array we store nested array of 1001 instead of 1000 to accomodate 10^6. To add a value we make the value true and to remove we make it false. 
//To check contains we return the value at that index.

class MyHashSet {

    private boolean[][] storage;
    private int buckets;
    private int bucketItems;

    public MyHashSet() {
        this.storage = new boolean[1000][];
        this.buckets = 1000;
        this.bucketItems = 1000;        
    }
    
    public int hash1(int key){
        return key%buckets;
    }
    public int hash2(int key){
        return key/bucketItems;
    }
    public void add(int key) {
        int bucket = hash1(key);
        int bucketItem = hash2(key);
        if(storage[bucket] == null) {
            if(bucket == 0){
                storage[bucket] = new boolean[bucketItems + 1]; 
            }
            else{
                storage[bucket] = new boolean[bucketItems];
            }
        }
        storage[bucket][bucketItem] = true;
    }
    
    public void remove(int key) {
        int bucket = hash1(key);
        int bucketItem = hash2(key);
        if(storage[bucket] == null) return;
        storage[bucket][bucketItem] = false;
        
    }
    
    public boolean contains(int key) {
        int bucket = hash1(key);
        int bucketItem = hash2(key);
        if(storage[bucket] == null) return false;
        return storage[bucket][bucketItem];
    }
}

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet obj = new MyHashSet();
 * obj.add(key);
 * obj.remove(key);
 * boolean param_3 = obj.contains(key);
 */

File: DesignHashSet.cs
 public class MyHashSet {
   //Approach: Determine size of primary and secondary array based on the dataset
   //          Determine the primary and secondary hash function
   //          If adding secondary array value for that key will be true and if removing secondary array value for that key will be false.


// Time Complexity : Add - O(N), Remove - O(N) and Contains - O(N)
// Space Complexity : Considering the large dataset O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
   
    private bool[][] storage;
        private int buckets;
        private int bucketItems;

        private int PrimaryHash(int key)
        {
            return key%buckets;
        }

        private int SecondaryHash(int key)
        {
            return key/bucketItems;
        }

        public MyHashSet()
        {
            buckets = 1000;
            bucketItems = 1000;
            storage = new bool[1001][];
        }

      
        public void Add(int key)
        {
            int bucket = PrimaryHash(key);
            if(storage[bucket] == null)
            {
                if (bucket == 0)
                {
                    storage[bucket] = new bool[bucketItems + 1];
                }
                else
                {
                    storage[bucket] = new bool[bucketItems];
                }
            }
            int bucketItem = SecondaryHash(key);
            storage[bucket][bucketItem] = true;
        }

        public void Remove(int key)
        {
            int bucket = PrimaryHash(key);
            if (storage[bucket] == null)
            {
                return;
            }
            int bucketItem = SecondaryHash(key);
            storage[bucket][bucketItem] = false;
        }

        public bool Contains(int key)
        {
            int bucket = PrimaryHash(key);
            if (storage[bucket] == null)
            {
                return false;
            }

            int bucketItem = SecondaryHash(key);
            return storage[bucket][bucketItem];
        }
}


